src/audio/pipewire.c

```diff
@@ -0,0 +1,254 @@
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    // stride is essentially how many samples there are per channel, in relation to how many channels there are in the output format
+    stride = sizeof(int16_t) * state->channel_count;   // not sure about this one, this could be wrong if the format is 8bit only. This is impractical nowadays, but speech dispatcher still has that case in some modules, so this is left as a mystery and a todo for later
+    number_of_frames = buf->datas[0].maxsize / stride; // how many frames of audio are in this chunk, aka how many channel_count touples are available this call to on_process. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases
+    if (b->requested > 0)
+    {
+        number_of_frames = SPA_MIN(number_of_frames, b->requested);
+    }
+    // if there's something to be read from the ring buffer at this time, we do so now
+    if (available_samples > 0)
+    {
+        available_for_loading = SPA_MIN(available_samples, number_of_frames);
+    }
+    else
+    {
+        available_for_loading = 0;
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = number_of_frames - available_for_loading;
+    if (available_for_loading > 0)
+    {
+        spa_ringbuffer_read_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, read_index, destination_memory, available_for_loading);
+        spa_ringbuffer_read_update(&state->rb, read_index + available_for_loading);
+    }
+    if (must_load_with_silence > 0)
+    {
+        memset(destination_memory + available_for_loading, 0, must_load_with_silence);
+    }
+    // now, we should have most of the data ready, we fill in some metadata and push the buffer object to pipewire
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->size = number_of_frames;
+    buf->datas[0].chunk->stride = stride;
+    pw_stream_queue_buffer(state->stream, b);
```

Comment:
You also need to notify pipewire_play, through a condition variable, to tell it that you have freed room in the sample_buffer, so it can fill it again.

Reply:
> as long as pipewire_play gets called every time there's a new chunk to be played, it will do the right thing, because it automatically fills the buffer with the data

No, it needs to block when the `frame_buffer` is full. And then needs to be woken up once it got some room.

> This would lead to data races if the ringbuffer implementation wasn't guarding access, but it does, so that's not an issue

It does protect access, yes. But it does not provide a blocking primitive. pipewire_play does need to block at some point, otherwise it doesn't know how fast it's supposed to push data to the `frame_buffer`.

Reply:
alright, finally, now I understand how this is supposed to work! figuring out how to solve those segmentation faults I'm seeing, then will be doing that

Reply:
fixed in latest commit, hopefully